packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js

```diff
@@ -3,278 +3,657 @@ import { FreehandLine as FreehandLineComponent } from '@plugins/drawingTools/com
-        const distFromEnd = self.points.length - lastPoint - 1
-        const distFromStart = firstPoint
-        spansStartPoint = (distFromEnd + distFromStart) < deleteCount
+    },
+
+    closeSplice() {
+      // remove our splice starting point 
+      self.splicePoints.shift()
+
+      let closeSpliceAction = {
+        type: 'splice-end',
+        pointIndexDrag: self.spliceDragPointIndex,
+        pointIndexClose: self.spliceClosePointIndex,
+        pointsFromDrag: [],
+        pointsFromClose: [],
+        pointsSpliceCount: self.splicePoints.length,
+        isReversed: self.spliceReverse,
+        throughBeginning: self.spliceThroughBeginning,
       }
-      if (!spansStartPoint) {
-        /*
-        Segment lies entirely within the line path, so splice points
-        from dragIndex to targetIndex.
-        */
-        self.splice(startIndex, endIndex)
+
+      let startIndex = (self.spliceReverse)
+        ? self.spliceClosePointIndex
+        : self.spliceDragPointIndex
+      let endIndex = (self.spliceReverse)
+        ? self.spliceDragPointIndex
+        : self.spliceClosePointIndex
+
+      // we don't want to remove the start index but everything in between
+      startIndex++
+
+      if (self.spliceThroughBeginning) {
+        // When we remove our spliced points we start the path at the lowest-indexed end
+        if (self.spliceReverse) self.splicePoints = self.splicePoints.reverse()
+        self.splicePoints.push(self.points.at(endIndex))
+
+        for (let i = startIndex; i < self.points.length; i++) {
+          closeSpliceAction.pointsFromClose.push(self.points.at(i))
+        }
+        for (let i = 0; i < endIndex; i++) {
+          closeSpliceAction.pointsFromDrag.push(self.points.at(i))
+        }
+
+        self.points.splice(startIndex, self.points.length - startIndex)
+        self.points.splice(0, endIndex)
+        self.points.splice(self.points.length, 0, ...self.splicePoints)
       } else {
-        /*
-        Segment spans the start point of a closed loop, so
-        trim the ends of the line back to dragIndex and targetIndex.
-        */
-        self.trim(startIndex, endIndex)
+        for (var i = startIndex; i < endIndex; i++) {
+          closeSpliceAction.pointsFromDrag.push(self.points[i])
+        }
+
+        if (self.spliceReverse) {
+          self.splicePoints = self.splicePoints.reverse()
+        }
+
+        self.points.splice(startIndex, endIndex - startIndex, ...self.splicePoints)
+      }
+
+      self.spliceReset()
+      self.setDragPoint(self.pathIsClosed ? null : self.points.at(-1))
+      self.setClosePoint(self.pathIsClosed ? null : self.points.at(0))
+
+      return closeSpliceAction
+    },
+
+    undo() {
+      if (self.drawActions.length == 0) {
+        // remove a mark that was created, not loaded
+        if (self.points.length == 1) {
+          self.finish()
+        }
+        return
+      }
+      let action = toJS(self.drawActions.at(-1))
```

Comment:
```suggestion
      const action = self.drawActions.at(-1)
```
`toJS` is a debugging tool. You probably want the observable `action` here.

You could also define a view that returns the last action on the `drawActions` stack, then reference that here.

Reply:
True, but this line works without the `toJS`. I’m not following why the action can’t be observable here. I think I’m missing something.

Reply:
I'd also note that a getter will be cached, and its value only computed when the underlying observable array changes.. We use that a lot elsewhere in the classifier eg. in the subject queue here:
https://github.com/zooniverse/front-end-monorepo/blob/9bad4eadce28314a889fe042b9e6622e2450d6d4/packages/lib-classifier/src/store/SubjectStore/SubjectStore.js#L58-L66